### PR TITLE
stuff

### PR DESCRIPTION
**Pull request by SylveonBottle**
Added:

- Jazz hands (it's funny)
- Automatic fake latency based on distance/melee weapon
- Something to randomize your ping (so you don't have constant solid 342 ping that never increases/decreases when using high fakelag)
- Lots of sliders to customize certain settings
- A lot more things will be toggled/switched to legit mode when spectated now (needs testing)
- A few "secret" features I started working on but gave up after a few hours

Updated:

- Increased leg jitter (2 wasn't enough when tested. Now it works while crouching. Spycrabing with leg jitter is the most hilarious thing I've seen recently)
- Changed a few names to be more consistant
- Changed position of a few things to be more consistant
- Disabled community-server-only feature.. that got you community-server banned..
- Made some randomized timers a little more consistant(?)

This could probably be cleaned up / optimized some more, but so far this was a very fun learning experience for me.

### CODE REVIEW

**Comment by SylveonBottle** [comment]
Things needed in order to improve the current code:
-figure out how to translate relative position into `pCmd:SetSideMove` for fast-stop
-figure out how to find if a person spectating you is set to rage/priority
-figure out what the condition effect for stun is (the [tf2 wiki for stunning](https://wiki.teamfortress.com/wiki/Stun) has no mention of what the tf_cond is, and the [tf2 wiki for tf_cond](https://wiki.teamfortress.com/wiki/Cheats) has no mention of stunning
-maybe re-add the first person for 'legit on spectate', maybe as an option
-figure out why removing sleep(0) breaks auto fakelatency (this scares me)

File: Misc-Tools.lua [hunk 1]
```diff
@@ -117,66 +161,102 @@ local function OnCreateMove(pCmd)
...
+    end
+
     -- Anti RTD
     if mRemovals:IsSelected("RTD Effects") then
         if CurrentRTD == "Cursed" then
             pCmd:SetForwardMove(pCmd:GetForwardMove() * (-1))
             pCmd:SetSideMove(pCmd:GetSideMove() * (-1))
         elseif CurrentRTD == "Drugged" or CurrentRTD == "Bad Sauce" then
-            SetOptionTemp("norecoil", 1)
+            --SetOptionTemp("norecoil", 1) --can get you banned in community servers 🤔
```
Comment (lnx00) [inline Misc-Tools.lua:192]: I think i should remove this whole RTD section anyways and only keep the screen effect removals

File: Misc-Tools.lua [hunk 2]
```diff
@@ -117,66 +161,102 @@ local function OnCreateMove(pCmd)
 
     -- Random Fakelag
     if mRandLag:GetValue() == true then
-        if pLocal:IsAlive() and pCmd.command_number % 20 == 0 then
-            local randValue = math.random(120, 315)
+        flTimer = flTimer +1
+        if (flTimer >= mRandLagValue:GetValue()) then
```
Comment (lnx00) [inline Misc-Tools.lua:165]: Why not use `pCmd.command_number % mRandLagValue:GetValue() == 0`?

File: Misc-Tools.lua [hunk 3]
```diff
@@ -117,66 +161,102 @@ local function OnCreateMove(pCmd)
...
+            flTimer = 0
+            local randValue = math.random(mRandLagMin:GetValue(), mRandLagMax:GetValue())
             gui.SetValue("fake lag value", randValue)
         end
     end
 
+    -- Random Ping
+    if mRandPing:GetValue() == true then
+        prTimer = prTimer +1
+        if (prTimer >= mRandPingValue:GetValue() * 66) then
```
Comment (lnx00) [inline Misc-Tools.lua:175]: Why not use `pCmd.command_number % (mRandPingValue:GetValue() * 66) == 0`?

File: Misc-Tools.lua [hunk 4]
```diff
@@ -18,35 +18,45 @@ local ObserverMode = {
...
-
+    }
 local Removals = {
     ["RTD Effects"] = false,
     ["HUD Texts"] = false
-}
+    }
+local prTimer = 0
+local flTimer = 0
+local ttTimer = 0
```
Comment (lnx00) [inline Misc-Tools.lua:28]: Unused

**Review by lnx00 — CHANGES_REQUESTED** [review]
Looks great so far, thanks! There are only some small changes that need to be done

**Comment by lnx00** [comment]
Regarding the `sleep(0)` issue: Is it possible that the sleep command cancels the execution of the function (since it throws an error)? You could try using `return` instead and see if this fixes the issue 🤔

File: Misc-Tools.lua (hunk 1 above)
Comment (SylveonBottle) [reply Misc-Tools.lua:192]: `return` works

File: Misc-Tools.lua [hunk 5]
```diff
@@ -91,18 +101,52 @@ local function OnCreateMove(pCmd)
...
+        if pCmd.command_number % 2 == 0 then
+            client.SetConVar("cl_flipviewmodels", 1 )
+        else
+            client.SetConVar("cl_flipviewmodels", 0)
+        end
+    end
+
+    -- Retry when stunned
+    -- if mRetryStunned:GetValue() == true then
+    --     if (pLocal:IsAlive()) and ????
```
[the remark this answers is not in the document]
Comment (SylveonBottle) [reply Misc-Tools.lua:136]: Thanks. Added, but do you know how you can check if the closest player to you is a heavy with the holiday punch?

File: Misc-Tools.lua [hunk 6]
```diff
@@ -117,66 +161,102 @@ local function OnCreateMove(pCmd)
...
     -- [[ Features that need to iterate through all players ]]
     local players = entities.FindByClass("CTFPlayer")
     for k, vPlayer in pairs(players) do
         if vPlayer:IsValid() == false then goto continue end
 
         -- Smooth on spectate
-        if mSpecSmooth:GetValue() == true then
-            local obsMode = pLocal:GetPropInt("m_iObserverMode")
+        if mLegitSpec:GetValue() == true then
+            --local obsMode = pLocal:GetPropInt("m_iObserverMode") -- todo: add first-person switch for first-person only (I don't want to rage-backstab people in third-person, it's still suspicious)
```
[the remark this answers is not in the document]
Comment (SylveonBottle) [reply Misc-Tools.lua:208]: I believe I've done this now.

File: Misc-Tools.lua (hunk 1 above)
Comment (SylveonBottle) [reply Misc-Tools.lua:192]: That's your call. Personally, I do appreciate the reverse direction keys.

File: Misc-Tools.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (lnx00) [reply Misc-Tools.lua:136]: Maybe you could use `m_hActiveWeapon`?

File: Misc-Tools.lua (hunk 2 above)
Comment (lnx00) [reply Misc-Tools.lua:170]: Hmm you're right. I just don't like having multiple variables only used for counting, but it's better than using command_number i guess

**Comment by SylveonBottle** [comment]
Added an automatic melee callout (I intend to add more). 
Added `Retry when stunned` (need help for further improvements, read the comment in the code)
Added a switch for firstperson only legit mode (needs testing to verify if it works).
Added a check to prevent auto latency auto melee from activating if you're cloaked.

File: Misc-Tools.lua (hunk 3 above)
Comment (SylveonBottle) [reply Misc-Tools.lua:180]: See https://github.com/lnx00/Lmaobox-LUA/pull/4#discussion_r894529633

**Review by SylveonBottle — COMMENTED** [review]
I don't know what button I pressed to "start a review" but this label next to my name is annoying. So ignore this comment.

**Review by lnx00 — APPROVED** [review]
Looks good so far. Is anything else coming or can this be merged?

**Comment by SylveonBottle** [comment]
You can go ahead and merge it now if you want, but I'll still be working on it for longer. I still plan on adding automations for voicelines.

In it's current state, it works fine. Although faststop is only in debug mode right now until I can figure out how to translate absolute velocity into relative velocity.

**Comment by SylveonBottle** [comment]
I don't know why, but this spits out an error.
`bad argument #1 to 'GetItemDefinitionByID' (number expected, got nil)`
```lua
    if vWeapon ~= nil then                                                         -- If "vWeapon" is not nil
        local vWeaponDefIndex = vWeapon:GetPropInt("m_iItemDefinitionIndex")       -- Set "vWeaponDefIndex" to the "vWeapon"'s item definition index
        local vWeaponDef      = itemschema.GetItemDefinitionByID(vWeaponDefIndex)  -- Set "vWeaponDef" to the local "vWeapon"'s item definition
        local vWeaponName     = vWeaponDef:GetName()                               -- Set "vWeaponName" to the local "vWeapon"'s actual name
    end
```

**Comment by lnx00** [comment]
It seems that it can't find the prop. Try using this netprop:
` "DT_EconEntity", "m_AttributeManager", "m_Item", "m_iItemDefinitionIndex"`

**Comment by SylveonBottle** [comment]
- Updated "legit on spectated" to not mess with values if they're not already on. (some values don't work for some reason)
- Added "Rocket Liens" (this is very glitchy, but impossible to fix thanks to `EstimateAbsVelocity()` being wonky)
- Added values for Auto Fake Latency to change when you're far/near the enemy.

When Fake Latency is on (for example, set to 400), my ping is consistently at 430. It never goes up, and never goes down. This is very suspicious. 
Because of that, I added Random Ping. This will toggle ping spoof on/off every X seconds. This way, your ping will drastically/gradually change from ~150-350, giving an illusion of inconsistent ping.
However, when using the previous Auto Fake Latency, your ping would occasionally touch 5ms when not near enemies. This rendered the original intent of Random Ping as useless.
Now, you can set Auto Fake Latency to 200 when away from enemies, and 500 when near enemies. With this, you can also use Random Ping in order for your ping to appear between ~100 - 500 instead of occasionally dipping down to 5ms.

There is still a small scenario where Random Ping will stay active for too long while not near an enemy. This may result in your ping dipping below ~30ms suspiciously.

Possible solutions:

1. Change Random Ping to not enable unless you have fake latency currently set above 200ms (maybe making this a sub-option for Auto Fake Latency compatibility)
2. Create a new module where Random Ping is used in order to attempt to make your ping an exact number that you want.

Although (1.) may be possible, (2.) would be an excellent feature to have. As far as I'm aware, Cathook (or was it fedoraware?) contains code that allows the user to change ping spoof in order to get a "desired ping".  However, lmaobox's code is hardcoded to lower your ping by 500ms with no way of changing it. And as far as I'm aware, the lmaobox dev refuses to acknowledge that this feature is even possible.

**Comment by lnx00** [comment]
> As far as I'm aware, Cathook (or was it fedoraware?) contains code that allows the user to change ping spoof in order to get a "desired ping". However, lmaobox's code is hardcoded to lower your ping by 500ms with no way of changing it. And as far as I'm aware, the lmaobox dev refuses to acknowledge that this feature is even possible.

Both Fedoraware and Cathook have this. It's really easy to implement by sending a specific `cl_cmdrate` to the server:
https://github.com/tf2cheater2013/Fedoraware/blob/5a23d95aca96005b454a9a9b75ba2ac9ae603c83/Fedoraware/TeamFortress2/TeamFortress2/Features/Misc/Misc.cpp#L215

**Comment by SylveonBottle** [comment]
> Btw, should I wait or can this already be merged?

You can merge this whenever you want. If you do, I'll just make a new pull request for any updates I add.
I'm sorta just using this as my own little testing stuff, learning how to write in lua.

**Pull request merged**